### PR TITLE
Extend Cmake for building hpx as distribution packages (refs #3575)

### LINKED ISSUE #3575: Merge openSUSE reproducible patch
https://bugzilla.novell.com/show_bug.cgi?id=1100677
patch here:
https://build.opensuse.org/request/show/653867

maybe needs an CMake option.

/cc @bmwiedemann

### PR DESCRIPTION
**Pull request by diehlpk**
Fixes #3575 

## Proposed Changes

  - Add HPX_WITH_BUILD_DISTRIBUTION, which disables rdsctp optimization. This option needs to be turned when we build hpx as an rpm package

## Any background context you want to provide?

Building the rpm packages Fedora's build infrastructure or openSuse's build infrastructure requires to disable some optimization, since we do not know of the target architecture provide it.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1055,6 +1055,13 @@ if((MSVC14 AND CMAKE_CL_64) OR ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang"))
 endif()
 
 ################################################################################
+# Some special handling of the compilation is need on build infrastructure for
+# generating packages for target architecture, see issue #3575
+
+hpx_option(HPX_WITH_BUILD_DISTRIBUTION BOOL
+  "Build HPX on the build infrastructure on any LINUX distribution (default: OFF)."
+  OFF ADVANCED)
+
```
Comment (hkaiser) [inline CMakeLists.txt]: Wouldn't this be better called `HPX_WITH_BUILD_RPM_DISTRIBUTION`?

Comment (junghans) [reply CMakeLists.txt]: There could be other non-rpm-based distribution, which want to do the same, e.g. Ubuntu.

Comment (sithhell) [reply CMakeLists.txt]: `HPX_WITH_BUILD_BINARY_PACKAGE`?

Comment (junghans) [reply CMakeLists.txt]: I like that

File: CMakeLists.txt [hunk 2]
```diff
@@ -1477,24 +1484,34 @@ else()
...
   endif()
 
+ endif()
+
   if(NOT HPX_WITH_RDTSC AND NOT HPX_WITH_RDTSCP)
     hpx_warn("Neither rdtsc nor rdtscp is available; some performance counters may report incorrect results")
   endif()
 endif()
 
+
```
Comment (bmwiedemann) [inline CMakeLists.txt]: nitpicking: There are some unrelated whitespace changes, which is slightly bad style.

**Comment by diehlpk** [comment]
> it is still "openSUSE", not openSuse (in the PR description ;-)
@junghans Could you update this? Thanks

File: CMakeLists.txt (hunk 2 above)
Comment (diehlpk) [reply CMakeLists.txt]: Removed

**Comment by junghans** [comment]
> > it is still "openSUSE", not openSuse (in the PR description ;-)
> > @junghans Could you update this? Thanks

@bmwiedemann, from your comment in https://github.com/STEllAR-GROUP/hpx/issues/3575#issuecomment-445208968 I thought "openSUSE" is the correct capitalization, right?

File: CMakeLists.txt [hunk 3]
```diff
@@ -1478,18 +1484,26 @@ else()
...
   # rdtsc is an x86 instruction that reads the value of a CPU time stamp
   # counter. rdtscp is an extension to rdtsc. The difference is that rdtscp is
   # a serializing instruction.
   hpx_cpuid("rdtsc" HPX_WITH_RDTSC
     DEFINITIONS HPX_HAVE_RDTSC)
 
+ # One can not assume if RDTSCP is available on the hardware
+ # of the build infrastructure, that it will be available on
+ # all potential target hardware, see Issue  #3575
+ if(NOT $HPX_WITH_BUILD_BINARY_PACKAGE)
```
Comment (msimberg) [inline CMakeLists.txt]: `HPX_WITH_BUILD_BINARY_PACKAGE` needs to be wrapped in curly braces. Could you also clean up the history? Thanks!

Comment (diehlpk) [reply CMakeLists.txt]: Added

**Comment by bmwiedemann** [comment]
@junghans yes, openSUSE is the correct way, but look at this PR's description...

**Comment by junghans** [comment]
I see @diehlpk can edit this!
